tests/server/job/test_job.py

```diff
@@ -0,0 +1,322 @@
+    runner2_tmp_path = tmp_path / "runner2"
+    runner2_tmp_path.mkdir()
+
+    backend_store_uri = f"sqlite:///{str(db_tmp_path / 'mlflow.db')}"
+
+    with _setup_job_runner(1, monkeypatch, runner1_tmp_path, backend_store_uri) as job_runner_proc:
+        job1_id = submit_job(basic_job_fun, {"x": 3, "y": 4, "sleep_secs": 0})
+        job2_id = submit_job(basic_job_fun, {"x": 5, "y": 6, "sleep_secs": 10})
+        job3_id = submit_job(basic_job_fun, {"x": 7, "y": 8, "sleep_secs": 0})
+        time.sleep(1)
+
+    # ensure the job runner process is killed.
+    job_runner_proc.wait()
+
+    with _setup_job_runner(1, monkeypatch, runner2_tmp_path, backend_store_uri):
+        # assert that job1 has done, job2 is running, and job3 is pending.
+        assert query_job(job1_id) == (JobStatus.DONE, 7)
+        assert query_job(job2_id) == (JobStatus.RUNNING, None)
+        assert query_job(job3_id) == (JobStatus.PENDING, None)
+        time.sleep(10)
+        # assert all jobs are done.
+        assert query_job(job1_id) == (JobStatus.DONE, 7)
+        assert query_job(job2_id) == (JobStatus.DONE, 11)
+        assert query_job(job3_id) == (JobStatus.DONE, 15)
+
+
+def test_job_queue_parallelism(monkeypatch, tmp_path):
+    # test job queue parallelism=2 and each job consumes 2 seconds.
+    with _setup_job_runner(2, monkeypatch, tmp_path):
+        job_ids = [submit_job(basic_job_fun, {"x": x, "y": 1, "sleep_secs": 2}) for x in range(4)]
+        time.sleep(2.5)
+
+        # assert that job1 and job2 are done, and job3 and job4 are running
+        assert query_job(job_ids[0]) == (JobStatus.DONE, 1)
+        assert query_job(job_ids[1]) == (JobStatus.DONE, 2)
+        assert query_job(job_ids[2]) == (JobStatus.RUNNING, None)
+        assert query_job(job_ids[3]) == (JobStatus.RUNNING, None)
+
+        time.sleep(2.5)
+        assert query_job(job_ids[2]) == (JobStatus.DONE, 3)
+        assert query_job(job_ids[3]) == (JobStatus.DONE, 4)
+
+
+def transient_err_fun(tmp_dir, succeed_on_nth_run):
+    """
+    This function will raise `TransientError` on the first (`succeed_on_nth_run` - 1) runs,
+    then return 100 on the `succeed_on_nth_run` run. The `tmp_dir` records the run state.
+    """
+    from mlflow.server.job import TransientError
+
+    if len(os.listdir(tmp_dir)) == succeed_on_nth_run:
+        return 100
+    with open(os.path.join(tmp_dir, uuid.uuid4().hex), "w") as f:
+        f.close()
+    raise TransientError(RuntimeError("test transient error."))
+
+
+def test_job_retry_on_transient_error(monkeypatch, tmp_path):
+    monkeypatch.setenv("MLFLOW_SERVER_JOB_TRANSIENT_ERROR_RETRY_BASE_DELAY", "1")
+    with _setup_job_runner(1, monkeypatch, tmp_path):
+        store = _get_job_store()
+
+        job1_tmp_path = tmp_path / "job1"
+        job1_tmp_path.mkdir()
+
+        job1_id = submit_job(
+            transient_err_fun, {"tmp_dir": str(job1_tmp_path), "succeed_on_nth_run": 4}
+        )
+        time.sleep(15)
```

Comment:
is there a way to avoid long sleep like this? this slows down tests.

Reply:
for this test, we have to use a long timeout, the Huey retry task will have some extra delay than the delay we set in the Retry config.
but only for this one test, so it should not be a big deal.